lib/core/src/Cardano/Pool/Ranking.hs

```diff
@@ -0,0 +1,206 @@
+-- +------------+-------------+------------------------------------------------+
+module Cardano.Pool.Ranking
+    (
+      -- * Formulas
+      desirability
+
+    , saturatedPoolRewards
+    , saturatedPoolSize
+
+      -- * Types
+    , EpochConstants (..)
+    , Pool (..)
+    , RelativeStakeOf (..)
+    , mkRelativeStake
+    , Lovelace (..)
+    , Margin
+    , unsafeMkMargin
+    , getMargin
+    , NonNegative (..)
+    , Positive (..)
+    , unsafeToPositive
+    , unsafeToNonNegative
+    )
+    where
+
+import Prelude
+
+import GHC.Generics
+    ( Generic )
+import GHC.TypeLits
+    ( Symbol )
+
+--------------------------------------------------------------------------------
+-- Formulas from spec
+--------------------------------------------------------------------------------
+
+-- | Non-myopic pool desirability according to section 5.6.1.
+--
+-- Is /not/ affected by oversaturation nor pool stake in general.
+desirability
+    :: EpochConstants
+    -> Pool
+    -> Double
+desirability constants pool
+    | f_saturated <= c = 0
+    | otherwise    = (f_saturated - c) * (1 - m)
+  where
+    f_saturated = saturatedPoolRewards constants pool
+    m = getMargin $ margin pool
+    c = getNonNegative $ getLovelace $ cost pool
+
+-- | Total rewards for a pool if it were saturated.
+--
+-- When a0 = 0 this reduces to just p*R*z0 (tested
+-- by @prop_saturatedPoolRewardsReduces@)
+saturatedPoolRewards :: EpochConstants -> Pool -> Double
+saturatedPoolRewards constants pool =
+    let
+        a0 = getNonNegative $ leaderStakeInfluence constants
+        z0 = unRelativeStake $ saturatedPoolSize constants
+        s = unRelativeStake $ leaderStake pool
+        _R = getNonNegative $ getLovelace $ totalRewards constants
+        p = getNonNegative $ recentAvgPerformance pool
+        -- ^ technically \hat{p} in the spec
+    in
+        (p * _R) / (1 + a0)
+        * (z0 + ((min s z0) * a0))
+
+-- | Determines z0, i.e 1 / k
+saturatedPoolSize :: EpochConstants -> RelativeStakeOf "pool"
+saturatedPoolSize constants =
+    RelativeStake $ 1 / fromIntegral (getPositive $ optimalNumberOfPools constants)
+
+--------------------------------------------------------------------------------
+-- Types
+--------------------------------------------------------------------------------
+
+data EpochConstants = EpochConstants
+    { leaderStakeInfluence :: NonNegative Double
+      -- ^ a_0
+    , optimalNumberOfPools :: Positive Int
+      -- ^ k
+    , totalRewards :: Lovelace
+      -- ^ Total rewards in an epoch. "R" in the spec.
+    } deriving (Show, Eq, Generic)
+
+data Pool = Pool
+    { leaderStake :: RelativeStakeOf "pool leader"
+    , cost :: Lovelace
+    , margin :: Margin
+    , recentAvgPerformance :: NonNegative Double
+      -- ^ An already averaged performance-value
+    } deriving (Show, Eq, Generic)
+
+newtype Lovelace = Lovelace { getLovelace :: NonNegative Double }
```

Comment:
Why are Lovelace represented as `Double` :thinking: ? These clearly are non-negative integral values?  Also, the extra `Lovelace` wrapping seems superfluous since we don't attach any properties to the fact that they are Lovelace apart from:

- The fact that they are non-negative
- The fact that they are integers

Which can be entirely captured with "NonNegative Integer", or, an existing type: "Natural".